# Blastocystis reference and placement tree

## Setup

Create Blastocystis database/tree build conda environment

```bash
conda env create environment.yml
```

Load conda environment

```bash
conda activate bf_its_blasto
```

To deactivate environment

```bash
conda deactivate bf_its_blasto
```

## 1. Generate reference database

```bash
git clone https://github.com/bfosso/ITSoneDB_qiime.git
cd ITSoneDB_qiime/
ls *gz | while read line; do gzip -d $line; done
```

Remove wordwrap from fasta file

```bash
awk '!/^>/ { printf "%s", $0; n = "\n" } /^>/ { print n $0; n = "" } END { printf "%s", n } ' 99_ref_ITSoneDB.fa > 99_ref_ITSoneDB.oneline.fa
```

Pull Blastocystis reads 

```bash
cd ..
grep "Blastocystis" ITSoneDB_qiime/99_ref_ITSoneDB.oneline.fa -A 1 | sed 's/--//' > blasto_ref.fa
```

Concatenate these with your Blastocystis ASVs and all those in NCBI nt (search details: ("Blastocystis"[Organism] OR blastocystis[All Fields]) AND internal[All Fields] AND transcribed[All Fields] AND spacer[All Fields]) to cover a wider range of potential subtypes

```bash
grep "Blastocystis" ../01-read_processing/rep_set.tax.txt | awk '{print $1}' | while read line; do grep -w $line ../01-read_processing/rep_set.fa -A 1 ; done > blasto_asvs.fa
cat blasto_asvs.fa blasto_ref.fa ncbi_nt_blasto.fasta > all_blasto.fa
```

Now can use this to pull more references (change -db to path to your locally installed nt blast database)

```bash
blastn -evalue 1e-10 -outfmt 6 -db /home/allie/refdb/nt/nt -query all_blasto.fa -out blast.out
```

Pull only those hits that are at least 25bp and 85% identity to account for high genetic diversity in Blastocystis and potential short ITS1 regions

```bash
awk -F"\t" '$3>=85.0 && $4>=25' blast.out > good.hits
```

Get unique subjects and coordinates 

```bash
awk -F"\t" '{print $2, "\t", $9, "\t", $10}' good.hits | sort | uniq > ncbi.query
```

Pull reads by genome coordinates

```bash
python3 ncbi_nuccore_coordinate_pull.py -i ncbi.query -s 1 -fc 2 -rc 3 > ncbi_hits.fa 
```

Only keep those that are annotated as Blastocystis

```bash
grep "Blastocystis" ncbi_hits.fa | awk '{print $1}' | sed 's/>//' > ncbi_blasto.ids
seqtk subseq ncbi_hits.fa ncbi_blasto.ids > ncbi_blasto.fa
```

## 2. Dereplicate reads

First concatenate ITSoneDB and ncbi sequences (no ASVs in reference tree)

```bash
cat ncbi_blasto.fa blasto_ref.fa > full_ref.fa
```

Dereplicate and cluster

```bash
vsearch --derep_fulllength full_ref.fa --output full_ref.uniq.fa 
vsearch --sortbylength full_ref.uniq.fa --output full_ref.sort.fa
vsearch --cluster_fast full_ref.sort.fa --centroids full_ref.clust.fa --id 0.99
```

## 3. Align sequences and build maximum likelihood tree

First need to fix headers for raxml

```bash
sed -i 's/:/-/' full_ref.clust.fa
```

Align reference sequences

```bash
mafft --auto full_ref.clust.fa > full_ref.align.fa
```

Trim alignment

```bash
trimal -in full_ref.align.fa -out full_ref.trimal.fa  -gt 0.3 -st 0.001 
```

Build tree (use reduced non redundant file generated by RAxML)

```bash
rm *ref.tre
raxmlHPC-PTHREADS -T 8 -m GTRCAT -c 25 -e 0.001 -p 31514 -f a -N 100 -x 02938 -n ref.tre -s full_ref.trimal.fa.reduced
```

## 4. Build placement tree with newly generated ITS sequences 

Generate annotations file




Generate taxonomy file for annotations

```bash
mkdir reference_trees
grep ">" all_entamoeba.fa | sed 's/>//' | sed 's/|/\t/' | sed 's/ /\t/' | sed 's/ /_/g' | sed 's/|/_/g' > annotations.ent.txt
mv annotations.ent.txt reference_trees/
```

Add header line in nano, then move reference tree to proper directory

```bash
cp RAxML_bipartitions.ref.tre entamoeba.tre 
mv entamoeba.tre reference_trees
```

### Get unassigned Entamoeba or nematodes

First we want to see if there are any reads that were unassigned that are actually Entamoeba or nematodes. You need to have a local copy of the ncbi nt database -- change your path in the command below to its location

```bash
mkdir placement_tree
cd placement_tree
blastn -query Dal_Afribiota_18S_V4_unresolved.fasta -out Dal_Afribiota_18S_V4_unresolved.blast.out -db ~/Desktop/referenceDB/ncbi_12.5.19/nt -perc_identity 0.99 -evalue 1e-10 -max_target_seqs 500 -outfmt 6
```

Get taxonomic assignments for each sequence using a LCA algorithm (here I used [MEGAN6 community edition](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004957))

Pull nematodes and Entamoeba assigned reads from your taxonomy results

```bash
grep "Entamoeba" Dal_Afribiota_18S_V4_unresolved.taxonomy.txt | awk '{print $1}' > ent.ids
grep "Nematoda" Dal_Afribiota_18S_V4_unresolved.taxonomy.txt | awk '{print $1}' > nem.ids
```

Now pull the reads by their ids

```bash
cat ent.ids | while read line; do grep -w $line Dal_Afribiota_18S_V4_unresolved.fasta -A 1; done > ent.fa
cat nem.ids | while read line; do grep -w $line Dal_Afribiota_18S_V4_unresolved.fasta -A 1; done > nem.fa
```

Now we need to modify ASV names and merge the Dal & MI datasets

```bash
sed -i 's/ASV/Dal_ASV/' Dal_Afribiota_18S_V4_*e.fasta
sed -i 's/ASV/MI_ASV/' MI_Afribiota_18S_V4_*fasta
sed -i 's/ASV/MI_ASV/' MI_Afribiota_18S_V4_ASV_table_20_no_plants_host.txt
sed -i 's/ASV/Dal_ASV/' Dal_Afribiota_18S_V4_sequence_table.18s_R1_lwp.txt
sed -i 's/ASV/Dal_ASV/' ent.fa
sed -i 's/ASV/Dal_ASV/' nem.fa
cat MI_Afribiota_18S_V4_entamoba_ASV.fasta Dal_Afribiota_18S_V4_entamoeba_ASV.fasta ent.fa > all_ent.fa
cat MI_Afribiota_18S_V4_nematode_ASV.fasta Dal_Afribiota_18S_V4_nematode.fasta nem.fa > all_nem.fa
```

Blast all Entamoeba reads to test for accuracy

```bash
blastn -query all_ent.fa -out all_ent.blast.out -db ~/Desktop/referenceDB/ncbi_12.5.19/nt -max_target_seqs 500 -outfmt 6
```

### EPA tree build

Align your (unaligned) sequences to your curated reference aligment (not the trimmed one)

```bash
cd placement_tree
sina -i ../all_entamoeba.align.fa --prealigned -o all_entamoeba.arb
sina -i all_ent.fa -r all_entamoeba.arb -o query.align.ent.fa --fs-msc 0.01 --fs-full-len=100
```

Concatenate your query sequences and reference sequences

```bash
cat query.align.ent.fa ../all_entamoeba.align.fa > queryPlus.align.ent.fa
```

Now can build your EPA tree

```bash
rm *ent.epa.tre 
raxmlHPC-PTHREADS-SSE3 -f v -G 0.2 -m GTRCAT -n ent.epa.tre -s queryPlus.align.ent.fa -t ../reference_trees/entamoeba.tre -T 2
```

Clean up tree so you can read into figtree

```bash
sed 's/QUERY___//g' RAxML_labelledTree.ent.epa.tre | sed 's/\[I[0-9]*\]//g' > RAxML_placementTree.ent.epa.tre
```

### Constraint tree build

As EPA tree does not have bootstrap information, also generate a constraint tree

```bash
raxmlHPC-PTHREADS-SSE3 -f a -N 100 -G 0.2 -m GTRCAT -n ent.cons.tre -s queryPlus.align.ent.fa -g ../reference_trees/entamoeba.tre -T 4 -x 25734 -p 25793
```

Root tree in figtree before next step

![entamoeba tree](entamoeba_tree.png)

Now you can use the output of your placement tree in the R script phyloseq_tree.r.ipynb


